libraries/chain/controller.cpp

```diff
@@ -1246,14 +1246,63 @@ struct controller_impl {
       }
    }
 
+   template<typename ForkDB, typename BSP>
+   void apply_irreversible_block(ForkDB& forkdb, const BSP& bsp) {
+      if (read_mode != db_read_mode::IRREVERSIBLE)
+         return;
+      controller::block_report br;
+      if constexpr (std::is_same_v<block_state_legacy_ptr, std::decay_t<decltype(bsp)>>) {
+         // before transition to savanna
+         apply_block(br, bsp, controller::block_status::complete, trx_meta_cache_lookup{});
+      } else {
+         assert(bsp->block);
+         if (bsp->block->is_proper_svnn_block()) {
+            // if chain_head is legacy, update to non-legacy chain_head, this is needed so that the correct block_state is created in apply_block
+            if (std::holds_alternative<block_state_legacy_ptr>(chain_head.internal())) {
+               auto prev = forkdb.get_block(bsp->previous(), include_root_t::yes);
+               assert(prev);
+               chain_head = block_handle{prev};
+            }
+            apply_block(br, bsp, controller::block_status::complete, trx_meta_cache_lookup{});
+         } else {
+            // only called during transition when not a proper savanna block
+            fork_db.apply_l<void>([&](const auto& forkdb_l) {
+               block_state_legacy_ptr legacy = forkdb_l.get_block(bsp->id());
+               fork_db.switch_to_legacy(); // apply block uses to know what types to create
+               fc::scoped_exit<std::function<void()>> e([&]{fork_db.switch_to_both();});
+               apply_block(br, legacy, controller::block_status::complete, trx_meta_cache_lookup{});
+               // irreversible apply was just done, calculate new_valid here instead of in transition_to_savanna()
+               assert(legacy->action_receipt_digests);
+               auto prev = forkdb.get_block(legacy->previous(), include_root_t::yes);
```

Comment:
you could use `fork_db_l` here, right?

Reply:
No, we need `fork_db_s` because `prev` needs to be a `block_state_ptr`.

Reply:
I didn't think it would return a  `block_state_ptr` because we just did `fork_db.switch_to(fork_database::in_use_t::legacy);`?

Reply:
`forkdb` here is the passed in `forkdb`. Not the `forkdb_l` from the lambda argument.